# Sequencing Source File Details

Poseidon 2.7.0 added an option to specify sequencing source data. This is a tab-separated table, much like the Janno file, but following [a different schema](https://github.com/poseidon-framework/poseidon2-schema/blob/master/sequencingSourceFile_columns.tsv). Note that the primary entities in this table are Sequencing entities (typically corresponding to DNA libraries or even multiple runs/lanes of the same library). The link to the Individuals listed in the Janno-file are made through a foreign-key relationship into `Poseidon_ID`.

Here is an example for such a file:

```
Poseidon_ID	sample_accession	study_accession	run_accession	sample_alias	secondary_sample_accession	first_public	last_updated	instrument_model	library_layout	library_source	instrument_platform	library_name	library_strategy	fastq_aspera	fastq_bytes	fastq_md5	fastq_ftp	read_count	submitted_ftp
Ash033.SG	SAMEA7050454	PRJEB39316	ERR4331996	2	ERS4811084	2021-04-12	2020-07-09	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash033_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/006/ERR4331996/ERR4331996.fastq.gz	649563861	9bd0fceb5ab46cb894ea33765c122e83	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/006/ERR4331996/ERR4331996.fastq.gz	23386349	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4331996/Ash033_all.merged.hs37d5.fa.cons.90perc.bam
Ash002.SG	SAMEA7050404	PRJEB39316	ERR4332592	1	ERS4811035	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash002_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/002/ERR4332592/ERR4332592.fastq.gz	194164761	6d8831f5bb8ba9870cb55f834e98ab4d	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/002/ERR4332592/ERR4332592.fastq.gz	6471092	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332592/Ash002_all.merged.hs37d5.fa.cons.90perc.bam
Ash040.SG	SAMEA7050455	PRJEB39316	ERR4332593	3	ERS4811085	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash040_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/003/ERR4332593/ERR4332593.fastq.gz	276693447	539852f3d7fb574b2a1e4f1c0059f163	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/003/ERR4332593/ERR4332593.fastq.gz	9442394	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332593/Ash040_all.merged.hs37d5.fa.cons.90perc.bam
Ash129.SG	SAMEA7050457	PRJEB39316	ERR4332594	5	ERS4811087	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash129_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/004/ERR4332594/ERR4332594.fastq.gz	2503643788	4991d0030214687c7c1706c640a2b37e	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/004/ERR4332594/ERR4332594.fastq.gz	76891816	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332594/Ash129_all.merged.hs37d5.fa.cons.90perc.bam
Ash131.SG	SAMEA7050458	PRJEB39316	ERR4332631	6	ERS4811088	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash131_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/001/ERR4332631/ERR4332631.fastq.gz	535696334	3c680096683defd93b35a5f4004e9e87	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/001/ERR4332631/ERR4332631.fastq.gz	16781594	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332631/Ash131_all.merged.hs37d5.fa.cons.90perc.bam
Ash136.SG	SAMEA7050460	PRJEB39316	ERR4332632	8	ERS4811090	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash136_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/002/ERR4332632/ERR4332632.fastq.gz	671099104	d37d7301c8459ff7b2e737e2ccb31582	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/002/ERR4332632/ERR4332632.fastq.gz	19835438	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332632/Ash136_all.merged.hs37d5.fa.cons.90perc.bam
Ash133.SG	SAMEA7050459	PRJEB39316	ERR4332633	7	ERS4811089	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash133_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/003/ERR4332633/ERR4332633.fastq.gz	2432326317	5e401ad70be16a2e95df5fb760499cee	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/003/ERR4332633/ERR4332633.fastq.gz	62454527	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332633/Ash133_all.merged.hs37d5.fa.cons.90perc.bam
cth217.SG	SAMEA7050474	PRJEB39316	ERR4332634	22	ERS4811104	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	cth217_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/004/ERR4332634/ERR4332634.fastq.gz	156091710	22f3a8fd26765dde15d045f3bf0f573a	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/004/ERR4332634/ERR4332634.fastq.gz	4481358	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332634/cth217_all.merged.hs37d5.fa.cons.90perc.bam
cth006.SG	SAMEA7050461	PRJEB39316	ERR4332635	9	ERS4811091	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	cth006_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/005/ERR4332635/ERR4332635.fastq.gz	505068759	685f3c0aafd9a74b1b71e6c14dae8f1c	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/005/ERR4332635/ERR4332635.fastq.gz	16351320	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332635/cth006_all.merged.hs37d5.fa.cons.90perc.bam
cth728.SG	SAMEA7050462	PRJEB39316	ERR4332636	10	ERS4811092	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	cth728_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/006/ERR4332636/ERR4332636.fastq.gz	221723438	3f23731850c9e48012bd71d4ccc773cd	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/006/ERR4332636/ERR4332636.fastq.gz	6209127	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332636/cth728_all.merged.hs37d5.fa.cons.90perc.bam
Ash128.SG	SAMEA7050456	PRJEB39316	ERR4332637	4	ERS4811086	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	Ash128_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/007/ERR4332637/ERR4332637.fastq.gz	9236424742	0ab64459acd809b0ae82c20ea82dacca	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/007/ERR4332637/ERR4332637.fastq.gz	302487376	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332637/Ash128_all.merged.hs37d5.fa.cons.90perc.bam
cth747.SG	SAMEA7050464	PRJEB39316	ERR4332638	12	ERS4811094	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	cth747_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/008/ERR4332638/ERR4332638.fastq.gz	604529786	a1a37e278ea0e4632e2a0d43adbaa79c	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/008/ERR4332638/ERR4332638.fastq.gz	18850253	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332638/cth747_all.merged.hs37d5.fa.cons.90perc.bam
cth739.SG	SAMEA7050473	PRJEB39316	ERR4332639	21	ERS4811103	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	cth739_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/009/ERR4332639/ERR4332639.fastq.gz	1315991788	424b8c541b55c1cbef84028ac8b6be70	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/009/ERR4332639/ERR4332639.fastq.gz	45491923	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332639/cth739_all.merged.hs37d5.fa.cons.90perc.bam
cth842.SG	SAMEA7050463	PRJEB39316	ERR4332640	11	ERS4811093	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	cth842_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/000/ERR4332640/ERR4332640.fastq.gz	577421074	76e8349b53f62dc04d59fc09f3f31278	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/000/ERR4332640/ERR4332640.fastq.gz	19618647	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332640/cth842_all.merged.hs37d5.fa.cons.90perc.bam
pch034.SG	SAMEA7050465	PRJEB39316	ERR4332641	13	ERS4811095	2021-04-12	2020-07-10	Illumina HiSeq 2500	SINGLE	GENOMIC	ILLUMINA	pch034_all	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/001/ERR4332641/ERR4332641.fastq.gz	240996196	27ecc647f2a82b5c1a7a36467f765d54	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/001/ERR4332641/ERR4332641.fastq.gz	6722479	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332641/pch034-b1e1l1p2_AGAACCG_L005_ARmerged.180302_ST-E00269_0258_BHJF7NCCXY.all.hs37d5.fa.cons.90perc.bam
CCH144.SG	SAMEA7050466	PRJEB39316	ERR4332643	14	ERS4811096	2021-04-12	2020-07-10	Illumina NovaSeq 6000	SINGLE	GENOMIC	ILLUMINA	CCH144_L1	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/003/ERR4332643/ERR4332643.fastq.gz	152189811	9d88cd959221c96105afbcb6bb0fb07f	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/003/ERR4332643/ERR4332643.fastq.gz	4212571	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332643/CCH144-b1e1l1p1_TACTTAG-GTCAACC_L001_ARmerged.200115_A00621_0169_BHYLNNDSXX.all.hs37d5.cons.90perc.bam
CCH163.SG	SAMEA7050467	PRJEB39316	ERR4332645	15	ERS4811097	2021-04-12	2020-07-10	Illumina NovaSeq 6000	SINGLE	GENOMIC	ILLUMINA	CCH163_L1	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/005/ERR4332645/ERR4332645.fastq.gz	86634518	513794058fdf06e160cad3579241fc19	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/005/ERR4332645/ERR4332645.fastq.gz	2428370	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332645/CCH163-b1e1l1p1_CGTAATC-ATGGTCT_L001_ARmerged.200115_A00621_0169_BHYLNNDSXX.all.hs37d5.cons.90perc.bam
CCH285.SG	SAMEA7050468	PRJEB39316	ERR4332646	16	ERS4811098	2021-04-12	2020-07-10	Illumina NovaSeq 6000	SINGLE	GENOMIC	ILLUMINA	CCH285_L1	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/006/ERR4332646/ERR4332646.fastq.gz	184093891	d3af0bb0802d2da7d68df3cabbeb4841	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/006/ERR4332646/ERR4332646.fastq.gz	4879361	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332646/CCH285-b1e1l1p1_GTTATAC-ACTTCAA_L001_ARmerged.200115_A00621_0169_BHYLNNDSXX.all.hs37d5.cons.90perc.bam
CCH289.SG	SAMEA7050469	PRJEB39316	ERR4332647	17	ERS4811099	2021-04-12	2020-07-10	Illumina NovaSeq 6000	SINGLE	GENOMIC	ILLUMINA	CCH289_L1	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/007/ERR4332647/ERR4332647.fastq.gz	179891674	a9d1316845f7c8dd6a3852a82cfc4c20	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/007/ERR4332647/ERR4332647.fastq.gz	5013829	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332647/CCH289-b1e1l1p1_CTTGCGG-CTCATGG_L001_ARmerged.200115_A00621_0169_BHYLNNDSXX.all.hs37d5.cons.90perc.bam
CCH290.SG	SAMEA7050470	PRJEB39316	ERR4332648	18	ERS4811100	2021-04-12	2020-07-10	Illumina NovaSeq 6000	SINGLE	GENOMIC	ILLUMINA	CCH290_L1	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/008/ERR4332648/ERR4332648.fastq.gz	30401831	352d76ddb75edee7e30ba0800940b126	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/008/ERR4332648/ERR4332648.fastq.gz	765274	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332648/CCH290-b1e1l1p1_AATCTTC-TGATAGT_L001_ARmerged.200115_A00621_0169_BHYLNNDSXX.all.hs37d5.cons.90perc.bam
CCH294.SG	SAMEA7050471	PRJEB39316	ERR4332649	19	ERS4811101	2021-04-12	2020-07-10	Illumina NovaSeq 6000	SINGLE	GENOMIC	ILLUMINA	CCH294_L1	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/009/ERR4332649/ERR4332649.fastq.gz	610643404	6a5689763b239ff60c24bc5313892c6e	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/009/ERR4332649/ERR4332649.fastq.gz	15356366	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332649/CCH294-b1e1l1p1_ACCAACG-ATTAAGG_L001_ARmerged.200115_A00621_0169_BHYLNNDSXX.all.hs37d5.cons.90perc.bam
CCH311.SG	SAMEA7050472	PRJEB39316	ERR4332650	20	ERS4811102	2021-04-12	2020-07-10	Illumina NovaSeq 6000	SINGLE	GENOMIC	ILLUMINA	CCH311_L1	WGS	fasp.sra.ebi.ac.uk:/vol1/fastq/ERR433/000/ERR4332650/ERR4332650.fastq.gz	308098723	b67b26467df5117d6269793a9a63349e	ftp.sra.ebi.ac.uk/vol1/fastq/ERR433/000/ERR4332650/ERR4332650.fastq.gz	7539845	ftp.sra.ebi.ac.uk/vol1/run/ERR433/ERR4332650/CCH311-b1e1l1p1_AACCATG-CTACTCT_L001_ARmerged.200115_A00621_0169_BHYLNNDSXX.all.hs37d5.cons.90perc.bam
```

To help with generating such a table from data at the [ENA](https://www.ebi.ac.uk/ena/browser/home), we have a [convenience script](https://github.com/poseidon-framework/scripts/blob/main/get_ena_table.py) which helps downloading the table, except for the Poseidon_ID column which needs to be manually added.

Note that due to the heterogenous way users submit data to these archives, many columns in this table are less formalised than they could be. We generally advice simply using the above script and copy whatever data is in the archives into the respective columns.
